File: src/Biome.py
import json, os, sys

from src.Rangerray import Rangerray

from src.constants import COLOUR_SUB_BIOMES, OCTAVE_COUNT
from src.functions import int_median

class Biome:

    def __init__(self, WORLD_NAME, parent_biome_name, name, config, noise_lower, noise_upper):
        self.parent_biome_name = parent_biome_name
        self.name = name
        self.full_name = self.parent_biome_name + "." + self.name
        self.noise_lower, self.noise_upper = noise_lower, noise_upper

        self.config = config[self.name]
        if COLOUR_SUB_BIOMES:
            self.colour = tuple(int_median([config["colour"], self.config["colour"]]))
        else:
            self.colour = tuple(config["colour"])

        self.persistence = float(self.config["persistence"])
        self.configure_values()

        self.height_displacement = int(self.config["height_displacement"])
        self.height_multiplier = float(self.config["height_multiplier"])
        self.altitude_surfaces = Rangerray(self.full_name, self.config["altitude_surfaces"])


    def configure_values(self):
        self.noise_scale = 0
        self.amplitudes = []
        amplitude = 1
        for i in range(OCTAVE_COUNT):
            self.amplitudes.append(amplitude)
            self.noise_scale += amplitude
            amplitude *= self.persistence

    def __str__(self):
        return self.parent_biome_name + "." + self.name

File: src/Biome.ts
import fs from "fs";
import Perlin from "./Perlin.js";
import Rangerray from "./Rangerray.js";
import SubBiome from "./SubBiome.js";
import World from "./World.js";
import { exitWithError } from "./functions.js";
import { checkBiomeConfig } from "./issue_checking.js";
import { loadJSON } from "./env_script.js";
import { BiomeConfig } from "./types.js";

class Biome {

    public name: string;
    public shortPath: string;
    public filepath: string;
    public rangerray: Rangerray<SubBiome>;
    public config: BiomeConfig;

    constructor(name: string, world: World) {
        this.name = name;
        this.shortPath = ["biomes", this.name + ".json"].join("/");
        this.filepath = [world.filepath, this.shortPath].join("/");

        if (!fs.existsSync(this.filepath)) exitWithError("Undefined biome", `An undefined biome named '${this.name}' is referenced.`, "CONFIG.json", "biomes");
        this.config = loadJSON<BiomeConfig>(this.filepath);

        checkBiomeConfig(this.name, this.config, world);

        world.biomeColours[this.name] = this.config["colour"].join(",");

        this.rangerray = this.createRangerray(this.config.ranges, world);
        
    }

    private createRangerray(ranges: [number, string][], world: World): Rangerray<SubBiome> {

        let noiseLower = 0;

        // TODO: clean up
        let output: [number, SubBiome][] = ranges.map((item: [number, string]) => {
            let noiseUpper = item[0];
            let subBiomeName = item[1];
            if (!Object.keys(this.config).includes(subBiomeName)) {
                exitWithError("Undefined sub-biome", `An undefined sub-biome named '${subBiomeName}' is referenced.`, this.shortPath, "ranges");
            }
            noiseUpper = Perlin.flatten(noiseUpper);
            let obj = new SubBiome(world, this, subBiomeName, this.config, noiseLower, noiseUpper);
            noiseLower = noiseUpper;
            return [noiseUpper, obj];
        });

        let rangerray = new Rangerray<SubBiome>(this.name, output);

        return rangerray;

    }

    public selectValue(height: number): SubBiome {
        return this.rangerray.selectValue(height);
    }

}

export default Biome;
